src/storage/txn/txn_status_cache.rs

```diff
@@ -0,0 +1,962 @@
+//! keys after prewritting, while many simple reads and writes executes
+//! frequently, thus these simple transactions frequently meets the lock left
+//! by the large transaction. It will be very inefficient for these small
+//! transactions to come back to the client and start resolve lock procedure.
+//! Even if the client side has the cache of that transaction, it still wastes
+//! an RTT.
+//!
+//! There would be more possibilities if we have such a cache in TiKV side: for
+//! read requests, it can check the cache to know whether it can read from the
+//! lock; and for write requests, if it finds the transaction of that lock is
+//! already committed, it can merge together the resolve-lock-committing and the
+//! write operation that the request needs to perform.
+
+use std::{
+    sync::{atomic::AtomicU64, Arc},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use crossbeam::utils::CachePadded;
+use parking_lot::Mutex;
+use tikv_util::{
+    lru,
+    lru::{GetTailEntry, LruCache},
+};
+use txn_types::TimeStamp;
+
+use crate::storage::metrics::*;
+
+const TXN_STATUS_CACHE_SLOTS: usize = 128;
+
+/// An cache item should be kept for at least this time.
+/// Actually this should be guaranteed only for committed transactions. See
+/// [this section](#
+/// for-filtering-out-unwanted-late-arrived-stale-prewrite-requests) for details
+/// about why this is needed.
+const CACHE_ITEMS_REQUIRED_KEEP_TIME: Duration = Duration::from_secs(30);
+
+struct CacheEntry {
+    commit_ts: TimeStamp,
+    /// The system timestamp in milliseconds when the entry is inserted to the
+    /// cache.
+    insert_time: u64,
+}
+
+/// Defines the policy to evict expired entries from the cache.
+/// [`TxnStatusCache`] needs to keep entries for a while, so the common
+/// policy that only limiting capacity is not proper to be used here.
+struct TxnStatusCacheEvictPolicy {
+    required_keep_time_millis: u64,
+    #[cfg(test)]
+    simulated_system_time: Option<Arc<AtomicU64>>,
+}
+
+impl TxnStatusCacheEvictPolicy {
+    fn new(
+        required_keep_time: Duration,
+        #[allow(unused_variables)] simulated_system_time: Option<Arc<AtomicU64>>,
+    ) -> Self {
+        Self {
+            required_keep_time_millis: required_keep_time.as_millis() as u64,
+            #[cfg(test)]
+            simulated_system_time,
+        }
+    }
+
+    #[inline]
+    #[cfg(not(test))]
+    fn now(&self) -> SystemTime {
+        SystemTime::now()
+    }
+
+    /// When used in tests, the system time can be simulated by controlling the
+    /// field `simulated_system_time`.
+    #[inline]
+    #[cfg(test)]
+    fn now(&self) -> SystemTime {
+        // Always get the system time to simulate the latency.
+        let now = SystemTime::now();
+        if let Some(pseudo_system_time) = &self.simulated_system_time {
+            UNIX_EPOCH
+                + std::time::Duration::from_millis(
+                    pseudo_system_time.load(std::sync::atomic::Ordering::Acquire),
```

Comment:
It seems `Relaxed` is enough here for the atomic opereation? Though it would not be used in product.

Reply:
I'm afraid that `Relaxed` would cause strange instruction reordering here. I want to ensure that later accesses will always get the latest value that was just updated.

Reply:
It seems the atomic variable is only set in the test, a little bit weird the interfaces does not use both `acquire` and `release` together. It's just a matter of personal habit, and it's okay. I just asked casually.

Reply:
🤔Release is used when setting it in the test code below